### DISCUSSION TITLE
SVG and PNG file interface confusingly different with transparency

### DISCUSSION THREAD

**Discussion by ilyagr** [ImageMagick/ImageMagick#7600] [Kudos and Rants]
Firstly, thanks for making the tool!

Extending https://github.com/ImageMagick/ImageMagick/discussions/7596#discussioncomment-10575050, I have no more actionable and clear questions, but I still have a rant left in me.

Let's start with two input files that look the same and are conceptually the same, and perform the same operations on them. This is ImageMagick 7.1.1-37, Mac OS X Sonoma on ARM, with `magick` using Inkscape using to rasterize the SVG.

| Command  | Result for `test.svg`  | Result for `test.png` |
|----------|------------|----------|
| `INPUT` (original files)    | ![test](https://github.com/user-attachments/assets/db82db63-046f-4701-8fa8-53830976caa6) | ![bgtr](https://github.com/user-attachments/assets/6010b134-9618-4128-984d-43eb4af8c125)
| `magick INPUT -flatten output.png` (this is a fine, non-confusing example for context) | ![noop](https://github.com/user-attachments/assets/78219cb4-350d-409d-88cd-598dee95ac91) | ![noop](https://github.com/user-attachments/assets/78219cb4-350d-409d-88cd-598dee95ac91) 
| `magick INPUT output.png` | ![noop](https://github.com/user-attachments/assets/78219cb4-350d-409d-88cd-598dee95ac91) | ![bgtr](https://github.com/user-attachments/assets/6010b134-9618-4128-984d-43eb4af8c125)
| `magick INPUT -background black output.png` | ![noop](https://github.com/user-attachments/assets/78219cb4-350d-409d-88cd-598dee95ac91) | ![bgtr](https://github.com/user-attachments/assets/6010b134-9618-4128-984d-43eb4af8c125)
| `magick INPUT -background black -flatten output.png` | ![noop](https://github.com/user-attachments/assets/78219cb4-350d-409d-88cd-598dee95ac91) | ![bgtr-bgbk-flatten](https://github.com/user-attachments/assets/7dd5b02f-b494-4eeb-bba0-d2864bcd98d0)
| `magick -background black INPUT output.png` | ![bgtr-bgbk-flatten](https://github.com/user-attachments/assets/7dd5b02f-b494-4eeb-bba0-d2864bcd98d0) | ![bgtr](https://github.com/user-attachments/assets/ae1d82da-94e0-4bb6-bfe1-03283d1d780e)
| `magick -background black INPUT -flatten output.png` (another fine, non-confusing example) | ![bgtr-bgbk-flatten](https://github.com/user-attachments/assets/7dd5b02f-b494-4eeb-bba0-d2864bcd98d0) | ![bgtr-bgbk-flatten](https://github.com/user-attachments/assets/7dd5b02f-b494-4eeb-bba0-d2864bcd98d0) 

For bonus points, how does one convert `test.svg` into `test.png`? (Nothing like it appears in the left column)

Answer: `magick -background none test.svg test.png`.

-------------

I hope this is considered as a something that is desirable to, eventually, fix. It seems that a lot of this would work more sensibly if the rasterization of SVGs did not come out "already flattened" conceptually.

(The flattening operation itself is a bit confusing to me, but this could probably be solved with some reading)

**Update/P.S.:** One thing that does reassuringly make sense, and seems to be quite useful to wrap one's mind around this behavior, is that `magick -background none test.svg <ANYTHING> out.png` does seem to give the same result as `magick test.png <ANYTHING> out.png`. For example, `magick -background none test.svg -background black -flatten out.png` works the same as one would expect if they learned `magick test.png -background black -flatten out.png`.

So, if `-background none test.svg` behavior were the default for SVG files, things would make sense.

Perhaps it should be mentioned somewhere in the docs of the SVG format that it is strongly recommended to use `-background none` before the input file if the SVG has transparency and if you want things to be consistent with how `magick` handles bitmap files with transparency.

**Comment by snibgo** [comment]
The documentation https://www.imagemagick.org/script/formats.php under the SVG heading could have a sentence added:
```
The rasterized SVG is drawn over the current "-background" setting, which is white by default.
```

From my reading of the SVG standard (eg https://www.w3.org/TR/compositing-1/#simplealphacompositingexamples ), the default rasterizing of SVG should be over a transparent background.

When IM rasterizes an SVG, it does so over a background defined by "`-background`", which is `white` by default, rather than `transparent`. Thus IM doesn't follow the SVG standard, which could be considered a bug.

If it is a bug, then how should it be fixed? Changing the default `-background` to `transparent` would fix the bug, but also change the behaviour of existing scripts that may not involve SVG files, so this would be a bad fix.

In my opinion, a better fix would be to change the rasterization process. Instead of rasterizing over the `-background` colour, it would rasterize over a new setting, called (for example) "`-svg-background`". The default value for this new setting would be "`transparent`" but could be changed to white or black or anything by an option in the command before reading the SVG.

Of course, this would change the behaviour of rasterizing SVG files, but would have no impact elsewhere.

**Reply by ilyagr** [reply]
Thanks for checking! I generally agree, but had a quick comment.

> In my opinion, a better fix would be to change the rasterization process.

+1. As you said, it would be a breaking change for SVG files, though.

>  Instead of rasterizing over the -background colour, it would rasterize over a new setting, called (for example) "-svg-background". The default value for this new setting would be "transparent" but could be changed to white or black or anything by an option in the command before reading the SVG.

This setting is not really necessary. You could force people to use `-background white -flatten` if they want a white background, just like with PNG files.

Or, there could be a setting that affects loading any file (not just SVG), something like `-load-with-background white`. It'd be nice to give it a name that makes it clear that it affects loading files, and should therefore be put in the command line before the file is loaded. The default would be `-load-with-background none`.

**Comment by urban-warrior** [comment]
Per @snibgo suggestion, we updated the documentation so alert that the default is a white background and how to instead change it to a transparent background.

**Reply by ilyagr** [reply]
Thank you very much!
